dropshot-auth/src/sig.rs

```diff
@@ -0,0 +1,170 @@
+use async_trait::async_trait;
+use digest::KeyInit;
+use dropshot::{
+    ApiEndpointBodyContentType, Extractor, ExtractorMetadata, HttpError, RequestContext, ServerContext, UntypedBody,
+};
+use hmac::Mac;
+use std::{borrow::Cow, marker::PhantomData, sync::Arc};
+
+use crate::{
+    http::{internal_error, unauthorized},
+    FromBytes,
+};
+
+/// A request body that has been verified by an HMAC verifier `T`.
+#[derive(Debug)]
+pub struct HmacVerifiedBody<T, BodyType> {
+    audit: HmacVerifiedBodyAudit<T, BodyType>,
+}
+
+impl<T, BodyType> HmacVerifiedBody<T, BodyType>
+where
+    BodyType: FromBytes<HttpError>,
+{
+    /// Attempts to deserialize the request body into the specified `BodyType`. Returns a
+    /// [`BAD_REQUEST`](http::status::StatusCode::BAD_REQUEST) [`HttpError`](dropshot::HttpError) if the deserialization of `BodyType` fails
+    #[allow(dead_code)]
+    pub fn into_inner(self) -> Result<BodyType, HttpError> {
+        self.audit.into_inner()
+    }
+}
+
+/// A request body that performs the HMAC verification specified by the verifier `T`, but does not
+/// fail extraction when verification fails. The [`HmacVerifiedBodyAudit`] can be queried to determine
+/// if verification failed.
+#[derive(Debug)]
+pub struct HmacVerifiedBodyAudit<T, BodyType> {
+    body: UntypedBody,
+    _body_type: PhantomData<BodyType>,
+    content_type: ApiEndpointBodyContentType,
+    verified: bool,
+    _verifier: PhantomData<T>,
+}
+
+impl<T, BodyType> HmacVerifiedBodyAudit<T, BodyType>
+where
+    BodyType: FromBytes<HttpError>,
+{
+    /// Returns that status of if this body passed verification
+    pub fn verified(&self) -> bool {
+        self.verified
+    }
+
+    /// Attempts to deserialize the request body into the specified `BodyType`. Returns a
+    /// [`BAD_REQUEST`](http::status::StatusCode::BAD_REQUEST) [`HttpError`](dropshot::HttpError) if the deserialization of `BodyType` fails.
+    pub fn into_inner(self) -> Result<BodyType, HttpError> {
+        BodyType::from_bytes(self.body.as_bytes(), &self.content_type)
+    }
+}
+
+/// A trait to be used to implement various HMAC verification strategies. By default a strategy
+/// must implement two functions, one to provide the secret to the verifier, and one to extract
+/// the signature to check from a request. Additionally, a strategy can implement a custom function
+/// for extracting the materials from a request that should be signed.
+#[async_trait]
+pub trait HmacSignatureVerifier {
+    type Algo: Mac + KeyInit;
+
+    /// Provides the key to be used in signature verification.
+    async fn key<'a, Context: ServerContext>(rqctx: &'a Arc<RequestContext<Context>>) -> anyhow::Result<Cow<'a, [u8]>>;
```

Comment:
I refactored to using a `Vec<u8>` for the `fn`s outside of the `content` function. This was an overeager attempt to reduce cloning data from the request. In practice trying to borrow the `key` or `signature` didn't really work out.

The `content` `fn` though does borrow its response from the `UntypedBody` so that we don't need to clone it. The exception being Slack where they require to prepend data from a header to the request body. At which point I don't think we can avoid copying the data since we need to keep the original body around for deserialization later.